### ISSUE TITLE
feature request: allow for external or internal plugins for commands

### ISSUE THREAD

**Issue by alexanderkjeldaas** [paul-gauthier/aider#99]
blue-sky idea, but I can see use-cases where plugins to aider is useful:

- attach a PR/MR to an issue in github/gitlab
- pull/select PR/MR that should be reviewed (by querying github/gitlab APIs)
- import gitlab/github issue description into context before starting to implement a feature.
- import hackmd document into context (could be a design document)

aider could provide a file-based API  (file x contains context, y contains list of edited files, z contains bar), and then exec aider-foo for the /foo command, similar to how `git` handles external commands.

**Comment by FallenChromium** [comment]
These features could potentially use `hub` / `gh` to ease the setup for the user

**Comment by paul-gauthier** [comment]
Thanks for making this suggestion. It's certainly an interesting idea. 

I think that this is probably out of scope for the project in the near term. Although, if you have specific integrations or even ideas for implementing a more general system, I'd be happy to discuss them in terms of a draft pull request, or something along those lines.

**Comment by paul-gauthier** [comment]
I'm going to close this issue for now, but feel free to re-open or file a new issue any time.

**Comment by arash-bizcover** [comment]
Why closing @paul-gauthier ?
This is a very reasonable and super useful feature, no?

**Comment by pcfreak30** [comment]
While I think plugins CAN make sense, it honestly also makes sense to just use the MCP standard that seems to be getting mindshare, which is effectively JSON RPC for AI.

**Comment by harleypig** [comment]
I have heard mentions of this, but I don't really know what MCP is (hopefully not Master Control Program) or how it could be used with Aider.

End of line.

**Comment by pcfreak30** [comment]
> I have heard mentions of this, but I don't really know what MCP is (hopefully not Master Control Program) or how it could be used with Aider.
> 
> End of line.

https://modelcontextprotocol.io/introduction

**Comment by harleypig** [comment]
So, if I understand this correctly, aider would still need to be modified to manage this. Is this a case of 6 of one or half a dozen of the other?

**Comment by pcfreak30** [comment]
> So, if I understand this correctly, aider would still need to be modified to manage this. Is this a case of 6 of one or half a dozen of the other?

no idea what that means, sorry.

**Comment by harleypig** [comment]
It means other than implementation, I don't see much difference between modifying aider to support a plugin system or mcp. In either case aider needs to be significantly modified.

**Comment by pcfreak30** [comment]
> It means other than implementation, I don't see much difference between modifying aider to support a plugin system or mcp. In either case aider needs to be significantly modified.

I don't think anyone was disputing that.

**Comment by adrianlzt** [comment]
Related with #3314 and #2525 .

I think MCP could be better, that a custom plugin solution, because you can reuse MCP servers already implemented.

They compare MCP to LSP (language server protocol). With LSP, you create one for golang and use it in vim, vscode, etc.

**Comment by harleypig** [comment]
I like that argument. It makes understanding the idea of MCP a little easier.
